server/api/public.py

```diff
@@ -42,15 +47,106 @@
 # Specifying election results in a query param could cause us to hit URL size limits
 @api.route("/public/sample-sizes", methods=["POST"])
 def public_compute_sample_sizes():
-    compute_sample_sizes_input = request.get_json()
-    validate(compute_sample_sizes_input, COMPUTE_SAMPLE_SIZES_INPUT_SCHEMA)
-    sample_sizes = {
-        "ballotComparison": {},
-        "ballotPolling": {},
-        "batchComparison": {},
+    contest_for_sampler = parse_compute_sample_sizes_input(request.get_json())
+
+    audit_types = ["ballotComparison", "ballotPolling", "batchComparison"]
+    sample_sizes = {audit_type: {} for audit_type in audit_types}
+    sample_size_functions = {
+        "ballotComparison": compute_ballot_comparison_sample_size,
+        "ballotPolling": compute_ballot_polling_sample_size,
+        "batchComparison": compute_batch_comparison_sample_size,
     }
-    for risk_limit_percentage in list(range(0, 21)):
-        sample_sizes["ballotComparison"][str(risk_limit_percentage)] = 0
-        sample_sizes["ballotPolling"][str(risk_limit_percentage)] = 0
-        sample_sizes["batchComparison"][str(risk_limit_percentage)] = 0
+    for audit_type in audit_types:
+        for risk_limit_percentage in list(range(0, 21)):
+            sample_size_function = sample_size_functions[audit_type]
+            sample_sizes[audit_type][str(risk_limit_percentage)] = sample_size_function(
+                contest_for_sampler, risk_limit_percentage
+            )
+
     return jsonify(sample_sizes)
+
+
+def parse_compute_sample_sizes_input(
+    compute_sample_sizes_input: Any,
+) -> sampler_contest.Contest:
+    validate(compute_sample_sizes_input, COMPUTE_SAMPLE_SIZES_INPUT_SCHEMA)
+
+    election_results = compute_sample_sizes_input["electionResults"]
+    candidates = election_results["candidates"]
+    num_winners = election_results["numWinners"]
+    total_ballots_cast = election_results["totalBallotsCast"]
+
+    if all(candidate["votes"] == 0 for candidate in candidates):
+        raise Conflict("At least 1 candidate must have greater than 0 votes")
+
+    if num_winners >= len(candidates):
+        raise Conflict("Number of winners must be less than number of candidates")
+
+    candidate_votes_sum = sum(candidate["votes"] for candidate in candidates)
+    if total_ballots_cast < candidate_votes_sum:
+        raise Conflict("Total ballots cast cannot be less than sum of candidate votes")
+
+    contest = Contest(
+        choices=[
+            ContestChoice(
+                id=f"candidate-{i}",
+                num_votes=candidate["votes"]
+                + 0.01  # Safeguard against math errors involving 0s
+                if candidate["votes"] == 0
```

Comment:
Happy to discuss this `+ 0.01`! The primary edge case that this is protecting against is the case that one candidate has 0 votes and the number of total ballots cast is higher than the winning candidate's votes, e.g.

```
Helga Hippo: 10 votes
Bobby Bear: 0 votes
Total ballots cast: 11 votes <-- Important that this is greater than Helga Hippo's votes
```

The ballot polling computation crashes with inputs like this because `s_w` is 1 so `z_l` is negative infinity.

```python
        else:
            z_w = (2 * s_w).ln()
            z_l = (2 - 2 * s_w).ln()
    
            T = Decimal(min(get_test_statistics(contest.margins, sample_results).values()))
    
            weighted_alpha = (Decimal(1.0) / alpha) / T
            return int(
                (
>                   (weighted_alpha.ln() + (z_w / Decimal(2))) / (p_w * z_w + p_l * z_l)
                ).quantize(Decimal(1), rounding=ROUND_CEILING)
            )
E           decimal.InvalidOperation: [<class 'decimal.InvalidOperation'>]

server/audit_math/bravo.py:73: InvalidOperation
```

When the number of total ballots cast exactly matches the winning candidate's votes, there's no issue because the landslide short-circuit kicks in:

https://github.com/votingworks/arlo/blob/e4b637ee5f96ed2cce7d266ec4f4792ee9434e65/server/audit_math/bravo.py#L358-L364

While I could change the ballot polling logic itself, I've been approaching this task conservatively and trying to avoid updates to the audit math directly.

Reply:
this is actually a really interesting corner case, nice find! We should probably update the Arlo backend to handle it (tho I defer to @jonahkagan)

Reply:
First off, I think the audit planner needs to have a way to handle computation errors, because as we've learned, it's very hard to catch every audit math edge case ahead of time. @arsalansufi I think you've already got this in place, but just checking.

Given that we've caught this one (nice job @arsalansufi!), I think we should try to fix it, but as a separate bug fix to the actual math. I don't think the audit planner should defend against it separately, because if it's a problem in the planner it's a problem in real audits as well.

Reply:
> I think the audit planner needs to have a way to handle computation errors

Confirming that the UI does handle unexpected errors!

![image](https://user-images.githubusercontent.com/12616928/192870009-48c0452e-7b5b-4286-a9b6-e1a519b3d03c.png)

> I think we should try to fix it, but as a separate bug fix to the actual math. I don't think the audit planner should defend against it separately, because if it's a problem in the planner it's a problem in real audits as well.

This line of reasoning makes perfect sense to me. To avoid scope creep on this PR, I've opened a follow-up bug task. I'm leaving the audit-planner-specific fix in place for now since, as a public page, I do think it's more likely to receive funky inputs and create PagerDuty noise for us. The follow-up ticket notes that we should remove the audit-planner-specific fix once the root fix is in place.

https://github.com/votingworks/arlo/issues/1601